### PR TITLE
Fix shardLeader cache not cleaned in connection error

### LINKED ISSUE #17392: [Bug]: Shard leader cache is not invalidated if search meet connection error
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 98e9527
- Deployment mode(standalone or cluster): both
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

When grpcClient failed to dial to address, an error will be returned without cleaning the cache.

### Expected Behavior

Do the doSearch without cache when connection error happens.

### Steps To Reproduce

_No response_

### Milvus Log

_No response_

### Anything else?

Might related to #17367 #17338 
Might related recovery issue of #17288

### PR DESCRIPTION
**Pull request by congqixia**
Fix #17392

/kind bug
/cc @XuanYang-cn 

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

### CODE REVIEW

File: internal/proxy/task_search.go [hunk 1]
```diff
@@ -415,13 +415,14 @@ func (t *searchTask) searchShard(ctx context.Context, leaders []nodeInfo, channe
 		}
 		result, err := qn.Search(ctx, req)
 		if err != nil {
-			return err
-		}
-		if result.GetStatus().GetErrorCode() == commonpb.ErrorCode_NotShardLeader {
 			log.Warn("QueryNode search returns error", zap.Int64("nodeID", nodeID),
 				zap.Error(err))
 			return errInvalidShardLeaders
```
Comment (yah01) [inline internal/proxy/task_search.go:420]: It shouldn't returns errInvalidShardLeaders if the err is context.Canceled, it is fine to return this if err is grpc error or context.DeadlineExceeded

Comment (congqixia) [reply internal/proxy/task_search.go:420]: Either way, proxy shall clean the cache and do search without cache.
Returning errInvalidShardLeaders is internal for proxy and just means: 
> Current shard leader cache is not working

Comment (congqixia) [reply internal/proxy/task_search.go:420]: If just return the err, proxy shall check grpc err, errInvalidShardLeaders, context.Cancelled,  context.DeadlineExceed, which is just to redundant.

File: internal/proxy/task_search.go [hunk 2]
```diff
@@ -302,7 +302,7 @@ func (t *searchTask) Execute(ctx context.Context) error {
 	}
 
 	err := executeSearch(WithCache)
-	if err == errInvalidShardLeaders || funcutil.IsGrpcErr(err) {
+	if err == errInvalidShardLeaders {
```
Comment (yah01) [inline internal/proxy/task_search.go:305]: It's better to check this with errors.Is(err, errInvalidShardLeaders) which also works if err is wrapped

**Pull request closed without merging**